pkg/ccl/backupccl/restore_planning.go

```diff
@@ -158,6 +161,242 @@ func synthesizePGTempSchema(
+	// ensure that the tables we create below won't conflict with anything in the
+	// backup.
+	if err := bumpDescIDSequenceKey(ctx, p, databasesByID, schemasByID, tablesByID, typesByID); err != nil {
+		return err
+	}
+
+	for _, table := range systemTablesWithConflictingIDs {
+		log.Infof(ctx, "moving system table %s to non-conflicting ID", table.GetName())
+		if err := moveConflictingSystemTable(ctx, p, table); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// moveConflictingSystemTable moves a system table to a higher ID.
+//
+// We do the move by creating new tables in the system database. Data from the
+// conflicting tables is moved to the temporary tables. The namespace and
+// descriptor entry for the conflicting table is removed, and then the new table
+// is renamed.
+func moveConflictingSystemTable(
+	ctx context.Context, p sql.PlanHookState, oldTable catalog.TableDescriptor,
+) error {
+	ctx, span := tracing.ChildSpan(ctx, "backupccl.moveConflictingSystemTable")
+	defer span.Finish()
+
+	codec := p.ExecCfg().Codec
+	return sql.DescsTxn(ctx, p.ExecCfg(), func(ctx context.Context, txn *kv.Txn, descsCol *descs.Collection) error {
+		// Create a new table with the same schema as the table we want to
+		// move.
+		newTable := oldTable.NewBuilder().BuildCreatedMutable().(*tabledesc.Mutable)
+
+		systemTableCopyName := fmt.Sprintf("crdb_internal_copy_%s", oldTable.GetName())
+		newTable.Name = systemTableCopyName
+
+		// Allow the upgrades package to dynamically assign an ID to the new system
+		// table.
+		newTable.ID = descpb.InvalidID
+		_, created, err := upgrades.CreateSystemTableInTxn(ctx, p.ExecCfg().DB, txn, codec, newTable)
+		if err != nil {
+			return err
+		}
+		if !created {
+			return errors.New("temporary table for system table already exists")
+		}
+
+		// Copy data from the old table to the new table.
+		insertStmt := fmt.Sprintf("INSERT INTO system.%s (SELECT * FROM system.%s)", newTable.GetName(), oldTable.GetName())
+		if _, err := p.ExecCfg().InternalExecutor.ExecEx(ctx, "copy-to-temp-table", txn,
+			sessiondata.InternalExecutorOverride{User: username.NodeUserName()}, insertStmt); err != nil {
+			return err
+		}
+
+		// Drop the old system table.
+		dropStmt := fmt.Sprintf("DROP TABLE system.%s", oldTable.GetName())
+		if _, err := p.ExecCfg().InternalExecutor.ExecEx(ctx, "drop-system-table", txn,
+			sessiondata.InternalExecutorOverride{User: username.NodeUserName()}, dropStmt); err != nil {
+			return err
+		}
+
+		// Rename the system table copy to the correct system table name.
+		renameStmt := fmt.Sprintf("ALTER TABLE system.%s RENAME TO system.%s", systemTableCopyName, oldTable.GetName())
+		if _, err := p.ExecCfg().InternalExecutor.ExecEx(ctx, "rename-system-table", txn,
+			sessiondata.InternalExecutorOverride{User: username.NodeUserName()}, renameStmt); err != nil {
+			return err
+		}
+
+		// ClearRange on the table span for the old system table that we've just migrated.
```

Comment:
Should we pull the rest of this block out of the txn, i.e. wait until after the copy/drop/rename/etc commits before we go and clear range since clear range is non-txn'l and thus would block a txn retry?

Reply:
good catch, done.